### PR TITLE
GP-822 Self-healing Loop

### PR DESCRIPTION
**Pull request by danielhedlund**
## What does this change?

Adds a "self-healing loop" to **media-remover**. That is, when having received a `media_not_required` message, but the item is not backed-up where it needs to be, this new functionality requests that such a copy be made. **media-remover** listens to notifications of these new copies, and re-checks if the _item-to-remove_ is now clear to actually delete.

## Why is the change needed?

We want to try and remove all `media_not_required` items, while still respecting the safeguards. This means we want to trigger some items to be backed-up, so that we can safely remove them from their current storage area.

## Important implementation details

`MediaNotRequiredMessageProcessor` has been heavily refactored and several classes have been extracted: `NearlineHelper`, `OnlineHelper`, and `PendingDeletionHelper`, and, some methods have been moved to the new processors `OnlineArchiveMessageProcessor` and `OnlineNearlineMessageProcessor`.

A `MediaRemoverMessageProcessor` has been added to **online-nearline**.

### CODE REVIEW

**Review by fredex42 — CHANGES_REQUESTED** [review]
Few small changes but generally great work as always

File: media_remover/src/main/scala/OnlineArchiveMessageProcessor.scala [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+                pendingDeletionRecordDAO.updateAttemptCount(rec.id.get, rec.attempt + 1)
+//                NOT_IMPL_outputDeepArchiveCopyRequiredForNearline(rec)
+                rec.vidispineItemId match {
+                  case Some(vsItemId) => outputDeepArchiveCopyRequiredForNearline(vsItemId, rec.mediaTier.toString, Some(rec.originalFilePath))
+                  case None => throw SilentDropMessage(Some(s"Cannot request deep archive copy for ${rec.mediaTier.toString}, ${rec.originalFilePath}, $nearlineId - missing vidispine ID!"))
+                }
+            })
+          case None =>
+            logger.warn(s"No nearline ID in pending deletion rec for media ${archivedRecord.originalFilePath}")
+            throw SilentDropMessage(Some(s"No nearline ID in pending deletion rec for media ${archivedRecord.originalFilePath}"))
```
[the remark this answers is not in the document]
Comment (danielhedlund) [reply media_remover/src/main/scala/OnlineArchiveMessageProcessor.scala:108]: Ah yes, makes sense.

File: media_remover/src/main/scala/helpers/NearlineHelper.scala [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+        logger.warn(s"Failed to remove nearline media oid=$nearlineId, path=$filepath, reason: ${exception.getMessage}")
+        Future(Left(s"Failed to remove nearline media oid=$nearlineId, path=$filepath, reason: ${exception.getMessage}"))
+    }
+
+  def dealWithAttFiles(vault: Vault, nearlineId: String, originalFilePath: String): Future[Either[String, String]] = {
+    val combinedRes = (for {
+      objectMatrixEntry <- callObjectMatrixEntryFromOID(vault, nearlineId)
+      proxyRes <- deleteSingleAttMedia(vault, objectMatrixEntry, "ATT_PROXY_OID", nearlineId, originalFilePath)
+      thumbRes <- deleteSingleAttMedia(vault, objectMatrixEntry, "ATT_THUMB_OID", nearlineId, originalFilePath)
+      metaRes <- deleteSingleAttMedia(vault, objectMatrixEntry, "ATT_META_OID", nearlineId, originalFilePath)
```
[the remark this answers is not in the document]
Comment (danielhedlund) [reply media_remover/src/main/scala/helpers/NearlineHelper.scala:207]: https://github.com/guardian/pluto-storagetier/blob/4d77ab463c790a656690f6084b36968240f2b5ef/online_nearline/src/main/scala/matrixstore/CustomMXSMetadata.scala#L88 would highly suggest that they are.

File: media_remover/src/test/scala/helpers/NearlineHelperSpec.scala [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+
+        val mockAssetFolderLookup = mock[AssetFolderLookup]
+
+        val vault = mock[Vault]
+        vault.getId returns "mockVault"
+
+        val filePath = "/path/to/some/file.ext"
+
+        val toTest = new NearlineHelper(mockAssetFolderLookup) {
+          override protected def callObjectMatrixEntryFromOID(vault: Vault, fileName: String): Future[ObjectMatrixEntry] = Future.fromTry(Failure(new RuntimeException("no oid for you")))
```
[the remark this answers is not in the document]
Comment (danielhedlund) [reply media_remover/src/test/scala/helpers/NearlineHelperSpec.scala:60]: Ahyes!

File: media_remover/src/main/scala/S3ObjectChecker.scala [hunk 4]
```diff
@@ -54,6 +55,27 @@ class S3ObjectChecker(client: S3AsyncClient, var bucketName: String)(implicit ec
...
      })
   }
+
+  def onlineMediaExistsInDeepArchive(checksumMaybe: Option[String], fileSize: Long, originalFilePath: String, objectKey: String): Future[Boolean] =
+    mediaExistsInDeepArchive(MediaTiers.ONLINE.toString, checksumMaybe, fileSize, originalFilePath, objectKey)
+
+ def nearlineMediaExistsInDeepArchive(checksumMaybe: Option[String], fileSize: Long, originalFilePath: String, objectKey: String): Future[Boolean] =
+    mediaExistsInDeepArchive(MediaTiers.NEARLINE.toString, checksumMaybe, fileSize, originalFilePath, objectKey)
+
+  private def mediaExistsInDeepArchive(mediaTier: String, checksumMaybe: Option[String], fileSize: Long, originalFilePath: String, objectKey: String): Future[Boolean] =
```
[the remark this answers is not in the document]
Comment (danielhedlund) [reply media_remover/src/main/scala/S3ObjectChecker.scala:65]: Right!

File: media_remover/src/test/scala/S3ObjectCheckerSpec.scala [hunk 5]
```diff
@@ -87,4 +89,69 @@ class S3ObjectCheckerSpec extends Specification with Mockito {
...
+////        implicit val mockChecksumChecker = mock[ChecksumChecker]
+//
+//        implicit val mockOnlineHelper = mock[helpers.OnlineHelper]
+//        implicit val mockNearlineHelper = mock[helpers.NearlineHelper]
+//        implicit val mockPendingDeletionHelper = mock[helpers.PendingDeletionHelper]
+//
+//        val assetFolderLookup = new AssetFolderLookup(fakeConfig)
+//
+//        val toTest = new MediaNotRequiredMessageProcessor(assetFolderLookup)
+//
```
[the remark this answers is not in the document]
Comment (danielhedlund) [reply media_remover/src/test/scala/S3ObjectCheckerSpec.scala:113]: Yeah

File: media_remover/src/main/scala/helpers/NearlineHelper.scala [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+          logger.info(s"Appliance reported checksum of $remoteChecksum for $oid on ${vault.getId}")
+          Future(Some(remoteChecksum))
+      })
+    } yield maybeMd5
+
+  def nearlineExistsInInternalArchive(vault: Vault, internalArchiveVault: Vault, nearlineId: String, filePath: String, fileSize: Long): Future[Boolean] = {
+    val filePathBack = putItBack(filePath)
+    for {
+      maybeChecksum <- getChecksumForNearline(vault, nearlineId)
+      //TODO add nearlineId to parameter list for logging purposes(?)
```
[the remark this answers is not in the document]
Comment (danielhedlund) [reply media_remover/src/main/scala/helpers/NearlineHelper.scala:154]: 😅

File: media_remover/src/main/scala/helpers/NearlineHelper.scala [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+          case Right(value) =>
+            logger.debug(value)
+            MessageProcessorConverters.futureEitherToMPRV(deleteMainNearlineMedia(vault, nearlineId, filepath, vidispineItemIdMaybe))
+        })
+      case (_, _, _) => throw new RuntimeException(s"Cannot delete from nearline, wrong media tier ($mediaTier), or missing nearline id ($nearlineIdMaybe)")
+    }
+
+
+  private def deleteMainNearlineMedia(vault: Vault, nearlineId: String, filepath: String, vidispineItemIdMaybe: Option[String]): Future[Either[String, Json]] =
+    // TODO do we need to wrap this with a Future.fromTry?
```
[the remark this answers is not in the document]
Comment (danielhedlund) [reply media_remover/src/main/scala/helpers/NearlineHelper.scala:190]: Alright

File: media_remover/src/main/scala/helpers/PendingDeletionHelper.scala [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+class PendingDeletionHelper (implicit
+                             ec: ExecutionContext,
+                             pendingDeletionRecordDAO: PendingDeletionRecordDAO) {
+
+  private val logger = LoggerFactory.getLogger(getClass)
+
+
+  def removeDeletionPendingByMessage(msg: OnlineOutputMessage): Future[Either[String, Int]] =
+    (msg.mediaTier, msg.vidispineItemId, msg.nearlineId) match {
```
[the remark this answers is not in the document]
Comment (danielhedlund) [reply media_remover/src/main/scala/helpers/PendingDeletionHelper.scala:22]: Yeah, should have been a different type in the `OnlineOutputMessage`

File: media_remover/src/main/scala/helpers/PendingDeletionHelper.scala [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+  def storeDeletionPending(msg: OnlineOutputMessage): Future[Either[String, Int]] =
+    msg.originalFilePath match {
+      case Some(filePath) =>
+        Try {
+          MediaTiers.withName(msg.mediaTier)
+        } match {
+          case Success(mediaTier) =>
+            val recordMaybeFut = mediaTier match {
+              case MediaTiers.ONLINE =>
+                pendingDeletionRecordDAO.findByOnlineIdForONLINE(Ensurer.validateNeededFields(Some(0), Some(""), msg.vidispineItemId)._3)
```
[the remark this answers is not in the document]
Comment (danielhedlund) [reply media_remover/src/main/scala/helpers/PendingDeletionHelper.scala:89]: Changing to `Ensurer.validateMediaId`

File: media_remover/src/main/scala/helpers/NearlineHelper.scala [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+   * @param fileSize Long representing the size of the file to match
+   * @return a Future, containing a sequence of ObjectMatrixEntries that match the given file path and size
+   */
+  //  def findMatchingFilesOnNearline_fromFileCopier(vault: Vault, filePath: Path, fileSize: Long) = {
+  def findMatchingFilesOnVault(mediaTier: MediaTiers.Value, vault: Vault, filePath: String, fileSize: Long): Future[Seq[ObjectMatrixEntry]] = {
+    logger.debug(s"Looking for files matching $mediaTier media with $filePath at size $fileSize on ${vault.getId}")
+    callFindByFilenameNew(vault, filePath)
+      .map(fileNameMatches => {
+        val nullSizes = fileNameMatches.map(_.maybeGetSize()).collect({ case None => None }).length
+        // TODO Decide if we need to/should do this? Flip it around maybe?
```
[the remark this answers is not in the document]
Comment (danielhedlund) [reply media_remover/src/main/scala/helpers/NearlineHelper.scala:94]: Alright, makes sense

File: media_remover/src/main/scala/OnlineArchiveMessageProcessor.scala [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+        }
+      case _ =>
+        logger.debug(s"ignoring archive confirmation, no pending deletion for this ${MediaTiers.NEARLINE} item with ${archivedRecord.originalFilePath}")
+        throw SilentDropMessage(Some(s"ignoring archive confirmation, no pending deletion for this ${MediaTiers.NEARLINE} item with ${archivedRecord.originalFilePath}"))
+    })
+  }
+
+
+  def handleDeepArchiveCompleteForOnline(archivedRecord: ArchivedRecord): Future[Either[String, MessageProcessorReturnValue]] = {
+    val (_, _, vsItemId) = Ensurer.validateNeededFields(Some(1), Some("path"), archivedRecord.vidispineItemId)
```
[the remark this answers is not in the document]
Comment (danielhedlund) [reply media_remover/src/main/scala/OnlineArchiveMessageProcessor.scala:118]: Will use `Ensurer.validateMediaId` instead

File: media_remover/src/main/scala/OnlineArchiveMessageProcessor.scala [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+    val itemNeedsArchiving = VidispineMediaIngested(List(VidispineField("itemId", itemId)))
+    Future(Right(MessageProcessorReturnValue(itemNeedsArchiving.asJson, Seq(RMQDestination("vidispine-events", "vidispine.itemneedsarchive.nearline"))))) 
+  }
+
+
+  def findPendingDeletionRecord(archivedRecord: ArchivedRecord): Future[Option[PendingDeletionRecord]] = {
+    (for {
+      byOriginal <- pendingDeletionRecordDAO.findBySourceFilenameAndMediaTier(archivedRecord.originalFilePath, MediaTiers.NEARLINE)
+      byUploaded <- pendingDeletionRecordDAO.findBySourceFilenameAndMediaTier(archivedRecord.uploadedPath, MediaTiers.NEARLINE)
+      // byVsItemId <- pendingDeletionRecordDAO.findByOnlineIdForNEARLINE(archivedRecord.vidispineItemId // maybe?)
```
[the remark this answers is not in the document]
Comment (danielhedlund) [reply media_remover/src/main/scala/OnlineArchiveMessageProcessor.scala]: Reworked this to get all `PendingDeletionRecord`s by `originalFilePath`, and check all until one is found, using `Future.find()` etc.

File: media_remover/src/test/scala/OnlineArchiveMessageProcessorSpec.scala [hunk 13]
```diff
@@ -0,0 +1,361 @@
...
+
+//      mockPendingDeletionRecordDAO.findAllByOriginalFilePathForNEARLINE(any) returns Future(Seq())
+      mockPendingDeletionRecordDAO.findAllByOriginalFilePathForNEARLINE(anOriginalFilePath) returns Future(Seq(pendingRec, pendingRec2))
+
+      mockNearlineHelper.getNearlineFileSize(any, any) returns Future(50L)
+      mockNearlineHelper.getChecksumForNearline(any, any) returns Future(Some(aChecksum))
+//      val m = mock[MediaRemovedMessage]
+//      val j = mock[Json]
+//        Future(Right(m.asJson))
+//        Future(Right(MediaRemovedMessage("NEARLINE", anOriginalFilePath, Some("VX-33"), Some("mxs-33")).asJson))
```
Comment (fredex42) [inline media_remover/src/test/scala/OnlineArchiveMessageProcessorSpec.scala:207]: Can we zap the commented lines?

File: media_remover/src/test/scala/OnlineArchiveMessageProcessorSpec.scala [hunk 14]
```diff
@@ -0,0 +1,361 @@
...
+        nearlineIdMaybe = any,
+        vidispineItemIdMaybe = eqTo(Some("VX-3")))
+
+      result must beRight
+//      val mediaRemovedMessage =
+//        result.right.get.content.as[MediaRemovedMessage].right.get must beAnInstanceOf[MediaRemovedMessage]
+//      mediaRemovedMessage.vidispineItemId must beSome("VX-3")
+//      mediaRemovedMessage.nearlineId must beSome("mxs-3")
+//      mediaRemovedMessage.originalFilePath mustEqual anOriginalFilePath
+//      mediaRemovedMessage.mediaTier mustEqual MediaTiers.NEARLINE.toString
```
Comment (fredex42) [inline media_remover/src/test/scala/OnlineArchiveMessageProcessorSpec.scala:234]: :zap:

**Review by fredex42 — CHANGES_REQUESTED** [review]
Apologies, couple more things to fix :(

File: media_remover/src/main/scala/S3ObjectChecker.scala [hunk 15]
```diff
@@ -54,6 +55,27 @@ class S3ObjectChecker(client: S3AsyncClient, var bucketName: String)(implicit ec
...
+    objectExistsWithSizeAndMaybeChecksum(objectKey, fileSize, checksumMaybe).map({
+      case true =>
+        logger.info(s"$mediaTier file with path $originalFilePath, objectKey $objectKey and size $fileSize exists, safe to delete from higher level")
+        true
+      case false =>
+        logger.info(s"$mediaTier file with path $originalFilePath: No file $objectKey with matching size $fileSize found, do not delete")
+        false
+    }).recover({
+      case err: Throwable =>
+        logger.warn(s"Could not connect to deep archive to check if copy of $mediaTier media exists, do not delete. Err: ${err.getMessage}")
```
[the remark this answers is not in the document]
Comment (danielhedlund) [reply media_remover/src/main/scala/S3ObjectChecker.scala:75]: Alrighty!

File: media_remover/src/main/scala/OnlineNearlineMessageProcessor.scala [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+              case Some(onlineSize) =>
+                val filePathBack = nearlineHelper.putItBack(rec.originalFilePath)
+                onlineExistsInVault(vault, vsItemId, filePathBack, onlineSize).flatMap({
+                  case true =>
+                    for {
+                      mediaRemovedMessage <- onlineHelper.deleteMediaFromOnline(rec)
+                      _ <- pendingDeletionRecordDAO.deleteRecord(rec)
+                    } yield mediaRemovedMessage
+                  case false =>
+                    pendingDeletionRecordDAO.updateAttemptCount(rec.id.get, rec.attempt + 1)
```
[the remark this answers is not in the document]
Comment (danielhedlund) [reply media_remover/src/main/scala/OnlineNearlineMessageProcessor.scala]: Added

File: media_remover/src/test/scala/OnlineArchiveMessageProcessorSpec.scala (hunk 13 above)
Comment (danielhedlund) [reply media_remover/src/test/scala/OnlineArchiveMessageProcessorSpec.scala]: Yup!

File: media_remover/src/test/scala/OnlineArchiveMessageProcessorSpec.scala (hunk 14 above)
Comment (danielhedlund) [reply media_remover/src/test/scala/OnlineArchiveMessageProcessorSpec.scala]: Begone!

**Review by fredex42 — APPROVED** [review]
And with that mammoth effort, we're done :partying_face: :tada:

**Pull request merged**